tls/s2n_ktls.c

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#ifdef S2N_PLATFORM_SUPPORTS_KTLS
+    /*
+     * https://github.com/aws/s2n-tls/issues/3813
+     *
+     * _GNU_SOURCE is needed for resolving the constant SOL_TCP
+     * when building `tls/s2n_ktls.c`.
+     */
+    #ifndef _GNU_SOURCE
+        #define _GNU_SOURCE
+        #include <netinet/tcp.h>
+        #undef _GNU_SOURCE
+    #else
+        #include <netinet/tcp.h>
+    #endif
+
+    #include <linux/tls.h>
+    #include <sys/socket.h>
+
+    #include "error/s2n_errno.h"
+    #include "tls/s2n_connection.h"
+    #include "utils/s2n_result.h"
+
+    #define S2N_TLS_ULP_NAME      "tls"
+    #define S2N_TLS_ULP_NAME_SIZE sizeof(S2N_TLS_ULP_NAME)
+#endif
+
+#include "tls/s2n_ktls.h"
+#include "utils/s2n_safety.h"
+#include "utils/s2n_socket.h"
+
+S2N_RESULT s2n_ktls_validate(struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(conn->config);
+    RESULT_ENSURE_REF(conn->secure);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite->record_alg);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite->record_alg->cipher);
+
+    /* TODO support TLS 1.3
+     *
+     * TLS 1.3 support requires sending the KeyUpdate message when the cryptographic
+     * KeyLimits are met. However, this is currently only possible by applying a
+     * kernel patch to support this functionality.
+     */
+    RESULT_ENSURE_EQ(conn->actual_protocol_version, S2N_TLS12);
+
+    /* Check if the cipher supports kTLS */
+    RESULT_ENSURE_EQ(conn->secure->cipher_suite->record_alg->cipher->ktls_supported, true);
+    /* TODO key length is a weak check. Is there a better mechanism? */
+    uint8_t key_size = conn->secure->cipher_suite->record_alg->cipher->key_material_size;
```

Comment:
This is a VERY weak check. You have zero guarantee that only AES_GCM_128 has the AES_GCM_128 key size. Why isn't ktls_supported sufficient?